로또 2단계 미션 제출합니다

iimpala opened the pull request:
다음 미션과 병행하느라 제출이 늦어졌습니다.
1단계 머지시 리뷰주신 부분 반영하고 수동 기능 추가했습니다!

dongjoon1251 reviewed (changes requested):
피드백을 많이 반영해주셨네요 :)
질문주신 부분에 대해서는 추가로 남긴 코멘트에 포함되어 있습니다.
몇가지 코드상에 남기지 못한 코멘트를 아래에 남깁니다. 확인 부탁드립니다.
- 상수로 선언한 부분은 항상 클래스 맨 위에 오게 맞춰주세요.
- toString() 메소드를 사용하지 않고, 출력에 관련된 로직은 출력을 담당하는 클래스로 모아주세요.
- List나 Map 생성시에 stream()을 통해서 생성되도록 리팩토링해보세요.
- 실패하는 테스트들도 추가해주시고 @ParameterizedTest를 통해서 다양한 값에 대해서 테스트가 수행되도록 추가해주세요 :)
- assert구문은 최대한 테스트 메소드 하나당 하나만 사용하도록 맞춰주세요.

dongjoon1251 reviewed (changes requested):
제가 드렸던 코멘트들이 일부 잘 반영되지 않은 것 같습니다. 
코멘트들이 중간에 생략 표시로 바로 확인이 안되어서 못 보신 부분이 있으셨던 걸로 예상됩니다.
코멘트가 잘 이해가지 않거나 리뷰이분께서 생각하시는 방향과 다르다면 의견 주시면 좋을 것 습니다.

tttuer commented:
> * toString() 메소드를 사용하지 않고, 출력에 관련된 로직은 출력을 담당하는 클래스로 모아주세요.

LottoGenerateStrategy 인터페이스는 삭제를 하고 다시 올렸습니다!
궁금한 점이 toString() 메소드를 사용하지 않았을때 각 클래스들에서 getter로 값을 가져와서 출력 부분에 사용해야 할 것이라 생각됩니다. 그런데 getter는 안쓰는 편이 좋다고 강의 중에 들었는데 어떻게 getter를 쓰지 않으면서 toString을 없앨 수 있을까요?

dongjoon1251 commented:
> > * toString() 메소드를 사용하지 않고, 출력에 관련된 로직은 출력을 담당하는 클래스로 모아주세요.
> 
> LottoGenerateStrategy 인터페이스는 삭제를 하고 다시 올렸습니다!
> 궁금한 점이 toString() 메소드를 사용하지 않았을때 각 클래스들에서 getter로 값을 가져와서 출력 부분에 사용해야 할 것이라 생각됩니다. 그런데 getter는 안쓰는 편이 좋다고 강의 중에 들었는데 어떻게 getter를 쓰지 않으면서 toString을 없앨 수 있을까요?

setter는 아예 없앨 수 있지만 getter를 아예 사용하지 않는 것은 사실 거의 불가능합니다. 대신 외부에서 필요한 필드들만 추려서 dto로 변환해서 전달하거나 외부에서 수정이 불가하도록 컬렉션을 immutable 한 컬랙션으로 래핑해서 전달할 수 있습니다. 그러면 getter로 전달받은 객체를 외부에서 수정할 수 없거나 수정해도 문제가 없을 수 있습니다. getter 사용을 지양하는 이유는 최대한 외부에서 클래스 내부의 값들을 변경하지 못하도록 막기 위해서 입니다. 답변이 도움이 되었으면 좋겠네요 :) 더 궁금하신점은 DM으로 바로바로 질문주셔도 됩니다!

dongjoon1251 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,49 +1,53 @@
+        for (int i = 0; i < manualLottoCount; i++) {
+            inputs.add(LottoInputView.inputManualLotto());
+        }
+        return inputs;
```
아래와 같은 형식으로 리팩토링해볼 수 있을 것 같습니다. 꼭 Stream으로 바꿔야한다는 규칙은 없지만 for 문을 stream으로 바꿔보고 가독성이 더 좋은 방향으로 앞으로 적용해보시면 좋을 것 같습니다.

return IntStream.range(0, manualLottoCount)
                .mapToObj(i-> LottoInputView.inputManualLotto())
                .collect(Collectors.toList())

tttuer replied:
컨트롤러에도 for문이 있는지는 제가 신경을 잘 못썼네요. 피드백 감사합니다

dongjoon1251 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,49 +1,53 @@
+        money = new Money(LottoInputView.askMoneyToBuyLotto());
+
+        int manualLottoCount = LottoInputView.askLottoPurchaseCount();
+        money.checkCanBuy(manualLottoCount);
```
money를 가지고 로또를 제대로 살수 있는지 검증하는 부분도 Lottos 내부로 이동이 가능해보입니다.

dongjoon1251 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,49 +1,53 @@
 
-    private void printLottoResults() {
-        lottoSimulatorView.printResult(lottoGame.getLottoResults(), lottoGame.profitPercentage());
+    private static void printBuyLottos(Lottos lottos) {
```
따로 메소드로 분리하지 않아도 무방해보입니다.

dongjoon1251 commented on src/main/java/controller/LottoController.java:
```diff
@@ -1,49 +1,53 @@
-        LottoController lottoController = new LottoController();
-        lottoController.startLottoSimulator();
+    private static AnswerLotto getAnswerLottoNumbers() {
+        String answerLottoNumbers = LottoInputView.askLottoNumberToAnswerLotto();
```
이 메소드도 따로 분리하지 않아도 무방해보입니다.

dongjoon1251 commented on src/main/java/domain/AnswerLotto.java:
```diff
@@ -27,13 +36,13 @@ private boolean isMatchedBonusNumber(List<LottoNumber> lotto) {
-                .filter(number -> answerLotto.contains(number))
+                .filter(answerLotto::contains)
                 .collect(Collectors.toList()).size();
     }
```
이부분은 아래와 같이 표현이 가능합니다.

return (int) lotto.stream()
                .filter(answerLotto::contains)
                .count();

dongjoon1251 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,57 @@
+        Map<LottoStatus, Integer> result = new HashMap<>();
+
+        Arrays.stream(LottoStatus.values())
+                .forEach(lottoStatus -> result.put(lottoStatus, 0));
```
아래와 같이 초기화가 가능합니다.
Map<LottoStatus, Integer> result = Arrays.stream(LottoStatus.values())
                .collect(Collectors.toMap(Function.identity(), lottoStatus -> 0));

dongjoon1251 commented on src/main/java/domain/LottoResult.java:
```diff
@@ -0,0 +1,57 @@
+                .forEach(lottoStatus -> result.put(lottoStatus, 0));
+
+        lottos.getLottos().stream()
+                .forEach(lotto -> result.put(answerLotto.getResult(lotto), result.get(answerLotto.getResult(lotto)) + 1));
```
이부분은 for문으로 표현하는것이 더 좋을수도 있고, 위의 Map 생성 스트림을 참고하셔서 리팩토링해보셔도 좋습니다.

dongjoon1251 commented on src/main/java/domain/Lottos.java:
```diff
@@ -1,27 +1,64 @@
+        inputs.stream()
+                .map(StringUtils::splitText)
+                .forEach(lotto -> lottos.add(new Lotto(generateManualLotto(lotto))));
+    }
```
아래와 같이 바꾸고 list를 합치거나 stream을 합치는 연산을 통해서 this.lottos에 할당해주는 것도 가능합니다.

private List<Lotto> generateManualLottos(List<String> inputs) {
        return inputs.stream()
                .map(StringUtils::splitText)
                .map(lotto -> new Lotto(generateManualLotto(lotto)))
                .collect(Collectors.toList());
    }

dongjoon1251 commented on src/main/java/domain/LottoStatus.java:
```diff
@@ -6,15 +6,16 @@
-    private final int matchedLottoNumberCount;
+    private final long winnings;
+    private final int matchingCount;
+    private final boolean isBonusNumberMatched;
```
필드가 사용되지 않는데 findStatus부분에서 사용해보면 어떨까요?

dongjoon1251 commented on src/main/java/domain/LottoStatus.java:
```diff
@@ -23,25 +24,27 @@
+                .findFirst()
                 .orElse(NONE);
     }
 
```
secondOrThird 와 같이 enum에 사용된 상수명을 이용한 함수명은 지양해주시기 바랍니다. SECOND, THIRD의 의미가 달라질 경우에 문제가 생긴것을 파악하기 힘듭니다.

dongjoon1251 commented on src/main/java/domain/LottoStatus.java:
```diff
@@ -23,25 +24,27 @@
+        return lottoStatuses.stream()
+                .filter(lotto -> lotto.getMatchingCount() == matchingCount)
+                .findFirst()
                 .orElse(NONE);
```
해당하는 matchCount, isBonusNumberMatched 가 없을때 NONE으로 맞추는 것이 좋을지 에러를 내는것이 더 나은 상황은 없을지 고민해보시기 바랍니다.

dongjoon1251 commented on src/main/java/domain/AnswerLotto.java:
```diff
@@ -27,13 +36,13 @@ private boolean isMatchedBonusNumber(List<LottoNumber> lotto) {
     private void checkLottoHasBonus(Lotto answerLotto, LottoNumber bonus) {
         if (answerLotto.contains(bonus)) {
-            throw new LottoException("로또번호에 보너스가 포함되어있습니다.");
+            throw new InvalidLottoException("로또번호에 보너스가 포함되어있습니다.");
```
예외 클래스 내부에 에러 메시지를 상수로 정의해서 사용하면 어떨까요?

dongjoon1251 commented on src/main/java/view/LottoOutputView.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public static void printResult(LottoResult lottoResult, long profit) {
+        System.out.println("당첨 통계\n-----------");
```
해당 문자열들도 상수로 분리해주면 어떨까요?

dongjoon1251 commented on src/test/java/domain/MoneyTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Test
+    void 로또_개수() {
```
한글로 메소드명을 작성해주셔도 괜찮지만 @DisplayName으로 모두 통일해보면 어떨까요?

dongjoon1251 reviewed (approved):
피드백을 잘 반영해주셨네요. 
아직 java8 stream 라이브러리에 익숙하지 않으신 것 같아서 참고하실만한 코멘트를 몇개 추가했습니다.
다른 코멘트들도 다른 미션 수행에 도움이 될것으로 생각되어 같이 남겼습니다.
이번 미션은 여기에서 마무리하도록 하겠습니다.
고생하셨습니다.

iimpala committed:
feat(LottoPurchaseBudget) 로또 비용차감 로직 작성

- 구입하려는 로또 개수를 받아 예산 안에 드는지 검증하는 로직 작성
- 로또 구입시 금액만큼 예산을 차감하는 로직 작성

iimpala committed:
feat(LottoTicketSeller) 수동 로또 구매 로직 작성

- 수동으로 로또 번호를 받아 구입하는 로직 추가

iimpala committed:
refactor(LottoPurchaseBudget): 에러 메시지 수정

공백 추가

iimpala committed:
refactor(Prize): Prize에서 view 의존성 제거

Prize클래스 안에 view에 대한 의존성이 있던 부분을 컨트롤러로 이동
- 로또의 결과를 String으로 변환하는 부분을 Prize에서 컨트롤러로 이동
- 로또 결과를 List<String>로 매핑하여 view에서 단순 출력만 할 수 있도록 수정

iimpala committed:
refactor(LottoNumber): CACHE 초기화 로직 수정

IntStream을 사용하여 캐시를 초기화 할 수 있도록 수정

iimpala committed:
refactor(LottoTicketSeller): 상수 추출

한 티켓에 들어가는 로또번호의 개수를 의미하는 숫자를 상수로 추출

iimpala committed:
refactor(LottoTicket): LottoTicket 리팩토링

- 상수와 인스턴스 변수 사이에 공백라인 추가
- 메소드명 변경 (compare -> countMatchedNumbers)

iimpala committed:
refactor(LottoResult): LottoResult 리팩토링

LottoResult가 일급 컬렉션의 성격을 띄도록 수정
- 결과 집계로직을 WinningLotto로 이전

Rok93 reviewed (changes requested):
동현님 2단계 구현 잘해주셨습니다. 👍💯
충분히 잘 구현해주셨지만 조금 더 짚고넘어갔으면하는 부분이있어서 코멘트 남겨두었습니다. 
확인해서 반영 부탁드릴게요. 🙏

현재 `iimpala` 브랜치로 PR 올려주셨는데, 이 PR은 닫아주시고 
step2 브랜치로 다시 PR 요청 부탁드릴게요. 🙏
(PR 코멘트에 이 PR 링크를 같이 달아주시면 좋을 것 같아요. 🙂)
